text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,481 @@
+    } else {
+        let semantically_meaningful_name_two = baz();
+        semantically_meaningful_name_two
+    }
+}
+```
+
+Rust-analyzer would give them feedback that they wrote extraneous code that could be reduced down,
+with a suggested fix.
+However, they feel the variable name is acting as a form of documentation and want to keep it.
+
+When the contributor checks for lints, they run
+```console
+$ cargo clippy
+   Compiling foo
+    Finished `foo` profile [optimized + debuginfo] target(s) in 15.85s
+```
+No lints reported.
+
+A non-rust-analyzer user could run the following to look for feedback:
+```console
+$ CARGO_BUILD_NITS=nit cargo clippy
+   Compiling foo
+note: creating a let-binding and then immediately returning it like `let x = expr; x` at the end of a block
+...
+    Finished `foo` profile [optimized + debuginfo] target(s) in 15.85s
+```
+
+They then go and post a PR.
+On the PR, the code above gets the following issue opened:
+```
+X Check failure
+
+Code scanning clippy
+
+creating a let-binding and then immediately returning it like `let x = expr; x` at the end of a block (warning)
+
+Show more details
+```
+This gives the contributor an opportunity to know of this potential improvement
+if they missed it and for the code reviewer to double check if they agreed with
+the author's determination to ignore it.
+
+For future PRs, Github should not report this as it should recognize that the report is for existing code.
+
+## Integrating with CI
+
+The above potential experience is dependent on being able to integrate linting with CI to report only "new lints" or "lints for changed code".
+
+For Github users, you can use [`clippy-sarif`](https://crates.io/crates/clippy-sarif) to report lints as SARIF alerts.
+Github [tracks these alerts across runs](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#providing-data-to-track-code-scanning-alerts-across-runs).
+
+These can also be implemented manually with any setup.
+For example, in a large, corporate, legacy C++ code base that [epage](https://github.com/epage/)
+worked with, a linter used a similar database to track what lints were "new" and reported those back to
+[ReviewBoard](https://www.reviewboard.org/).
+
+## Choosing lint levels
+
+When creating a lint or overriding a default lint level,
+its important to consider the semantics, something like:
+
+`error` is for hard errors that even prevent reporting of errors further in the build, running of tests, etc.
+This should be reserved for when there is no point building or testing anything further because of how broken the code is.
+
+`warn` generally has the semantics of a soft-error.
+Its non-blocking locally but will be blocked when the code is merged upstream.
+This should be used when eventual correctness or consistency is desired.
+
+`nit` is for coding suggestions for users that may or may not improve the code or may not need to be done yet.
```

Comment:
I find this definition of the levels is quite vague.  Also, as @Manishearth [notes](https://github.com/rust-lang/rfcs/pull/3730#issuecomment-2489677425), clippy and rusttc have very different interpretations of lint levels already.

I find the example of `clippy::needless_borrow` illuminating.  In projects where I'm in charge of the lint policy, I always allow that lint.  The needless borrow is sometimes helpful at the call site to signal to the reader that a borrow is occurring.  Also, during development and refactoring, the types change back and forth, and the `&` becomes alternately necessary, and redundant.  IME this lint leads to needless textual churn.

Similar arguments could be made about many (many most) lints in clippy's `complexity` category.  That lint category likes to normalise the code as much as possible, subject to preserving the meaning of the current code to the compiler.  But that *doesn't* necessarily preserve meaning of this code fragment in a slightly different context (ie, after future changes), nor the meanings implied to human readers.

Are we proposing to downgrade `clippy::complexity` to `nit`, then?  If not, why not?